src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs

```diff
@@ -2904,116 +2816,135 @@ private static bool IsReallyAType(TypeSymbol type)
                 (type.SpecialType != SpecialType.System_Void);
         }
 
-        private struct Candidates
+        private static void GetAllCandidates(Dictionary<TypeSymbolWithAnnotations, TypeSymbolWithAnnotations> candidates, ArrayBuilder<TypeSymbolWithAnnotations> builder)
         {
-            // The keys are types, compared ignoring dynamic, tuple names, and nullability.
-            // The values are different nullabilities associated with that type. Both keys and
-            // values are TypeSymbolWithAnnotations, although we could use bool[]? for values
-            // and recreate the TypeSymbolWithAnnotations from the key+value as needed.
-            private readonly Dictionary<TypeSymbolWithAnnotations, List<TypeSymbolWithAnnotations>> _dictionary;
-            private readonly EqualsIgnoringDynamicAndTupleNamesComparer _valueComparer;
-
-            internal Candidates(bool includeNullability)
-            {
-                _dictionary = new Dictionary<TypeSymbolWithAnnotations, List<TypeSymbolWithAnnotations>>(
-                    EqualsIgnoringDynamicAndTupleNamesComparer.WithoutNullability);
-                _valueComparer = includeNullability ?
-                    EqualsIgnoringDynamicAndTupleNamesComparer.WithNullability :
-                    EqualsIgnoringDynamicAndTupleNamesComparer.WithoutNullability;
-            }
+            builder.AddRange(candidates.Values);
+        }
 
-            internal void GetAll(ArrayBuilder<TypeSymbolWithAnnotations> builder)
+        private static void AddAllCandidates(
+            Dictionary<TypeSymbolWithAnnotations, TypeSymbolWithAnnotations> candidates,
+            HashSet<TypeSymbolWithAnnotations> bounds,
+            VarianceKind variance,
+            ConversionsBase conversions,
+            ref bool hadNullabilityMismatch)
+        {
+            foreach (var candidate in bounds)
             {
-                foreach (var values in _dictionary.Values)
+                var type = candidate;
+                if (!conversions.IncludeNullability)
                 {
-                    builder.AddRange(values);
+                    type = type.SetUnknownNullabilityForReferenceTypes();
                 }
```

Comment:
I think we don't want to be doing this if we want to preserve nullability that does not depend on flow analysis.  (To be addressed outside this PR)